File: capacity-club-api/src/domain-modules/workout/model/filter/workout.filter.ts
import { ApiProperty } from '@nestjs/swagger';
import { IsString, Length, IsOptional } from 'class-validator';
import { HomeTraining } from 'domain-modules/home-training/model/entity/home-training.entity';
import { IndoorTraining } from 'domain-modules/indoor-training/model/entity/indoor-training.entity';
import { TrainingCircuit } from 'domain-modules/training-circuit/model/entity/training-circuit.entity';

/**
 * Class used to filter workout.
 * This class defines the criteria used to filter a list of workout.
 */
export class WorkoutFilter {
  @ApiProperty()
  @IsString()
  @Length(1, 50)
  title: string;
  @ApiProperty()
  @IsOptional()
  training_circuits: TrainingCircuit[];
  @ApiProperty()
  @IsOptional()
  home_training: HomeTraining;
  @ApiProperty()
  @IsOptional()
  indoor_training: IndoorTraining;
}

File: capacity_club_mobile_app/lib/domain-features/data/workout/model/filter/workout_filter.dart
import 'package:capacity_club_mobile_app/core/model/extension/json_map_extension.dart';
import 'package:capacity_club_mobile_app/domain-features/data/home-training/model/home_training_model.dart';
import 'package:capacity_club_mobile_app/domain-features/data/indoor-training/model/indoor_training_model.dart';
import 'package:capacity_club_mobile_app/domain-features/data/training-circuit/model/training_circuit_model.dart';

class WorkoutFilter {
  final String? title;
  final List<TrainingCircuitModel>? training_circuits;
  final IndoorTrainingModel? indoor_training;
  final HomeTrainingModel? home_training;

  const WorkoutFilter({
    this.title,
    this.training_circuits,
    this.indoor_training,
    this.home_training,
  });

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> json = {};

    json.addIfNotNull('title', title);
    json.addIfNotNull(
      'training_circuits',
      training_circuits?.map((e) => e.toJson()).toList(),
    );
    json.addIfNotNull('indoor_training', indoor_training?.toJson());
    json.addIfNotNull('home_training', home_training?.toJson());

    return json;
  }
}
